### PR TITLE
3.0 stable & Price Model Abstraction

### PR DESCRIPTION
**Pull request by jspizziri**
I've upgraded the gem to 3.0 stable.

Also, I've taken the liberty to do a bit of data model abstraction, I created a `VolumePriceModel` which is associated with a role and a set of `VolumePrices`.

The point is to allow for several pricing models per variant and associate them with a role so that different pricing models can be applied for different roles i.e. wholesale vs. retail. This is described in more detail here: https://github.com/spree-contrib/spree_volume_pricing/issues/72

The specs aren't passing right now because of the data model refactor. Eventually I'll be able to get to those, assuming merging this in is desired.

### CODE REVIEW

**Comment by jspizziri** [comment]
@JDutil could you take at this?

There are significant structural changes. If you don't want to make these changes thats fine, but if you do then I can spend some more time to get the specs passing again.

**Comment by JDutil** [comment]
@jspizziri overall looks okay but builds pretty broken check the travis logs looks like you maybe missing an association

**Comment by jspizziri** [comment]
For sure. I'll clean it up, I primarily just wanted to make sure the architectural changes were kosher. I'll notify you when I'm done tidying up the build.

**Comment by juancroca** [comment]
Is there any update on this I would really appreciate is this was live

**Comment by jspizziri** [comment]
@juancroca, I'm currently developing on it on my fork, and actually it is useable (I'm currently using it in a production environment). The only thing that remains at this point is that I'm basically rewriting the spec's as they are pretty much obsolete now. 

I just pushed [my most recent changes](https://github.com/jspizziri/spree_volume_pricing/tree/3-0-stable) on my 3-0-stable branch in which I've got everything prepped for writing new specs. If you would like to help it would go faster. You could do so by forking my 3-0-stable branch, work on some model specs, and submit a PR to my fork.

Once the specs are rewritten and green I'll bug @JDutil again about getting it merged.

**Comment by jspizziri** [comment]
Also, I forgot to mention. It is very important that if this feature persists, to get it merged into this repository. Please, please, please, do not use my branch, but rather help me do work to get this merged in here. I don't want two warring branches of this extension.

**Comment by juancroca** [comment]
@jspizziri I would like with the tests and make this merge happen as fast as possible but would like to point out some stuff:

Im not sure if I'm not getting it right but the Variant model doesn't have the volume_prices association any more the volume_price_models has, know we have to send a user and then it will find the volume_prices based on the user's role. But in the variant_decorator.rb on lines 51, 54, 74 and 77 you are doing self. volume_prices which should throw an error since it doesn't have that association any more. Am I missing something?

File: Gemfile [hunk 1]
```diff
@@ -1,6 +1,7 @@
 source 'https://rubygems.org'
 
 gem 'guard-rspec', require: false
-gem 'spree', github: 'spree/spree', branch: 'master'
+gem 'spree', github: 'spree/spree', branch: '3-0-stable'
```
Comment (futhr) [inline Gemfile]: This is `master` branch.

File: spree_volume_pricing.gemspec [hunk 2]
```diff
@@ -15,14 +15,16 @@ Gem::Specification.new do |s|
   s.require_path = 'lib'
   s.requirements << 'none'
 
-  s.add_dependency('spree_core', '~> 2.4.0.beta')
+  s.add_dependency('spree_core', '~> 3.0.0.beta')
```
Comment (futhr) [inline spree_volume_pricing.gemspec]: `3.1.0.beta` for this branch.

File: Gemfile (hunk 1 above)
Comment (jspizziri) [inline Gemfile]: You're right in that it should be `master`. This was a proposal initially which is still being worked on.

Also, the `3-0-stable` branch is not nonexistent: https://github.com/spree/spree/tree/3-0-stable

**Comment by jspizziri** [comment]
@JDutil and @futhr,

Now that this spec is passing, can we get it reviewed and then create a `3-0-stable` branch out of it?

Once its in `3-0-stable` I'll update the specs as @futhr noted so that they are appropriate for master and submit a PR.

**Comment by JDutil** [comment]
Looks like @futhr created and updated a 3-0-stable branch already.  I've just updated master to use his 3-0-stable work, and bumped to 3.1.0.beta so this will need to be rebased against it.

**Comment by jspizziri** [comment]
@JDutil, I've rebased and opened another PR for the `3-0-stable` branch. Please review.

https://github.com/spree-contrib/spree_volume_pricing/pull/76

**Pull request closed without merging**